internal/controller/supervisorconfig/activedirectoryupstreamwatcher/active_directory_upstream_watcher.go

```diff
@@ -0,0 +1,341 @@
+	// - the sAMAccountType is for a normal user account.
+	defaultActiveDirectoryUserSearchFilter = "(&(objectClass=person)(!(objectClass=computer))(!(showInAdvancedViewOnly=TRUE))(|(sAMAccountName={})(mail={}))(sAMAccountType=805306368))"
+
+	// - is a group.
+	// - has a member that matches the DN of the user we successfully logged in as.
+	// - perform nested group search by default.
+	defaultActiveDirectoryGroupSearchFilter = "(&(objectClass=group)(member:1.2.840.113556.1.4.1941:={}))"
+)
+
+type activeDirectoryUpstreamGenericLDAPImpl struct {
+	activeDirectoryIdentityProvider v1alpha1.ActiveDirectoryIdentityProvider
+}
+
+func (g *activeDirectoryUpstreamGenericLDAPImpl) Spec() upstreamwatchers.UpstreamGenericLDAPSpec {
+	return &activeDirectoryUpstreamGenericLDAPSpec{g.activeDirectoryIdentityProvider}
+}
+
+func (g *activeDirectoryUpstreamGenericLDAPImpl) Namespace() string {
+	return g.activeDirectoryIdentityProvider.Namespace
+}
+
+func (g *activeDirectoryUpstreamGenericLDAPImpl) Name() string {
+	return g.activeDirectoryIdentityProvider.Name
+}
+
+func (g *activeDirectoryUpstreamGenericLDAPImpl) Generation() int64 {
+	return g.activeDirectoryIdentityProvider.Generation
+}
+
+func (g *activeDirectoryUpstreamGenericLDAPImpl) Status() upstreamwatchers.UpstreamGenericLDAPStatus {
+	return &activeDirectoryUpstreamGenericLDAPStatus{g.activeDirectoryIdentityProvider}
+}
+
+type activeDirectoryUpstreamGenericLDAPSpec struct {
+	activeDirectoryIdentityProvider v1alpha1.ActiveDirectoryIdentityProvider
+}
+
+func (s *activeDirectoryUpstreamGenericLDAPSpec) Host() string {
+	return s.activeDirectoryIdentityProvider.Spec.Host
+}
+
+func (s *activeDirectoryUpstreamGenericLDAPSpec) TLSSpec() *v1alpha1.TLSSpec {
+	return s.activeDirectoryIdentityProvider.Spec.TLS
+}
+
+func (s *activeDirectoryUpstreamGenericLDAPSpec) BindSecretName() string {
+	return s.activeDirectoryIdentityProvider.Spec.Bind.SecretName
+}
+
+func (s *activeDirectoryUpstreamGenericLDAPSpec) UserSearch() upstreamwatchers.UpstreamGenericLDAPUserSearch {
+	return &activeDirectoryUpstreamGenericLDAPUserSearch{s.activeDirectoryIdentityProvider.Spec.UserSearch}
+}
+
+func (s *activeDirectoryUpstreamGenericLDAPSpec) GroupSearch() upstreamwatchers.UpstreamGenericLDAPGroupSearch {
+	return &activeDirectoryUpstreamGenericLDAPGroupSearch{s.activeDirectoryIdentityProvider.Spec.GroupSearch}
+}
+
+func (s *activeDirectoryUpstreamGenericLDAPSpec) DetectAndSetSearchBase(ctx context.Context, config *upstreamldap.ProviderConfig) *v1alpha1.Condition {
+	config.GroupSearch.Base = s.activeDirectoryIdentityProvider.Spec.GroupSearch.Base
+	config.UserSearch.Base = s.activeDirectoryIdentityProvider.Spec.UserSearch.Base
+	if config.GroupSearch.Base != "" && config.UserSearch.Base != "" {
+		// Both were already set in spec so just return; no need to query the RootDSE
+		return &v1alpha1.Condition{
+			Type:    "SearchBaseFound",
+			Status:  v1alpha1.ConditionTrue,
+			Reason:  "Success",
```

Comment:
The production code should probably use string constants for the Types and Reasons on all of these Conditions.

Also, perhaps we should make a new Reason string to indicate that we did not try to autodetect the search base because it is going to use the search base that you configured, to differentiate it from the successful auto-detection case? Maybe something like "UsingConfigurationFromSpec"?

Reply:
Resolved (I think). I took your suggestion of "UsingConfigurationFromSpec", and also made the error reason string slightly more specific, "ErrorFetchingSearchBase" rather than just "Error". I think keeping "Success" for the successful query still makes sense, even though it's less descriptive. It seems consistent with the other Conditions.